usr/bs_hyc.c

```diff
@@ -0,0 +1,388 @@
+		return scsi_get_in_buffer(cmdp);
+	case WRITE:
+	case WRITE_SAME_OP:
+		return scsi_get_out_buffer(cmdp);
+	}
+}
+
+static int bs_hyc_cmd_submit(struct scsi_cmd *cmdp)
+{
+	struct scsi_lu     *lup = NULL;
+	struct bs_hyc_info *infop = NULL;
+	io_type_t           op;
+	size_t              length;
+	uint64_t            offset;
+	char               *bufp = NULL;
+	struct hyc_cmd     *hcmdp = NULL;
+	RequestID           reqid = kInvalidRequestID;
+
+	lup = cmdp->dev;
+	infop = BS_HYC_I(lup);
+
+	assert(infop->vmdk != kInvalidVmdkHandle);
+
+	op = scsi_cmd_operation(cmdp);
+	if (op == UNKNOWN) {
+		return -EINVAL;
+	}
+
+	hcmdp = hyc_cmd_alloc(infop, cmdp, op);
+	if (!hcmdp) {
+		return -ENOMEM;
+	}
+
+	offset = scsi_cmd_offset(cmdp);
+	length = scsi_cmd_length(cmdp);
+
+	bufp = scsi_cmd_buffer(cmdp);
+	set_cmd_async(cmdp);
+
+	pthread_mutex_lock(&infop->lock);
+	DLL_ADD(&infop->sched_cmd_list, &hcmdp->list);
+	pthread_mutex_unlock(&infop->lock);
+
+	switch (op) {
+	case READ:
+		reqid = ScheduleRead(infop->vmdk, hcmdp, bufp, length, offset);
+		break;
+	case WRITE:
+		reqid = ScheduleWrite(infop->vmdk, hcmdp, bufp, length, offset);
+		break;
+	case WRITE_SAME_OP:
+		/** TODO */
+	case UNKNOWN:
+	default:
+		assert(0);
+	}
+
+	/* If we got reqid, set it in hyc_cmd */
+	if (reqid != kInvalidRequestID) {
+		hcmdp->reqid = reqid;
+	} else {
+		eprintf("request submission got err invalid request\n");
+		pthread_mutex_lock(&infop->lock);
+		DLL_REM(&hcmdp->list);
+		pthread_mutex_unlock(&infop->lock);
+		target_cmd_io_done(hcmdp->cmdp, SAM_STAT_CHECK_CONDITION);
+		free(hcmdp);
+	}
+
+	return 0;
+}
+
+static void bs_hyc_handle_completion(int fd, int events, void *datap)
+{
+	struct bs_hyc_info   *infop;
+	uint64_t              c;
+	struct RequestResult *resultsp = NULL;
+	bool                  has_more = true; /** True for the first cmd completion */
+	int                   rc;
+	int                   i;
+
+	assert(datap);
+	infop = datap;
+
+	resultsp = infop->request_resultsp;
+
+	rc = eventfd_read(fd, &c);
+	assert(rc == 0);
+
+	while (has_more == true) {
+		uint32_t nr_results = GetCompleteRequests(infop->vmdk, resultsp,
+						infop->nr_results, &has_more);
+		pthread_mutex_lock(&infop->lock);
+		/* Process completed request commands */
+		for (i = 0; i < nr_results; ++i) {
+			struct hyc_cmd *hcmdp = (struct hyc_cmd *)resultsp[i].privatep;
+			assert(hcmdp);
+
+			assert(resultsp[i].result == 0);
+			target_cmd_io_done(hcmdp->cmdp, SAM_STAT_GOOD);
+
+			DLL_REM(&hcmdp->list);
+			free(hcmdp);
+		}
+		pthread_mutex_unlock(&infop->lock);
+		memset(resultsp, 0, sizeof (*resultsp) * infop->nr_results);
+	}
+}
+
+static int bs_hyc_open(struct scsi_lu *lup, char *pathp,
+			int *fdp, uint64_t *sizep)
+{
+	struct bs_hyc_info *infop = BS_HYC_I(lup);
+	int                 rc = 0;
+	int                 ffd = -1;
+	int                 efd = -1;
+	uint32_t            blksize;
+
+	ffd = backed_file_open(pathp, O_RDWR | O_LARGEFILE | O_DIRECT, sizep,
```

Comment:
this ffd is never closed. resource leak.

Reply:
done.